### ISSUE TITLE
PWM fan with EEB

### ISSUE THREAD

**Issue by ruggb** [MarlinFirmware/Marlin#2104]
See #2046 - all the history is there.

The fan works on pin 11 with hardware PWM, but the heaters stopped working.

The fan hardware PWM works with BOARD_RAMPS_EFB on pin 9, so why does changing it to RAMPS_EEB and changing FAN_PIN to pin 11 break it?

> The only thing changing EEB, EFB,,,,, should do is redirect the i/o pins, right??? (aside from whatever has to be done with 2 extruders). So why does redirecting the fan pin with EEB set not work? EEB should automatically redirect the fan and since there is the extender option pin 11 or 6 would be a good choice without servos. AND what is the purpose of FAN_SOFT_PWM option - if it is a FAN option why does it break the heaters??

**Comment by thinkyhead** [comment]
I wonder if pin 11 is being used for something else... Do you have any servos?

**Comment by ruggb** [comment]
Scott -- let's keep 2046 open all the history is there.

**Comment by ruggb** [comment]
If servos r defined 11 is used there. I have no servos and did not chg anything there.
If there is another definition for 11, why does it work when FAN_SOFT_PWM is defined?

the ONLY place I find a write command to fan_pin exists is here in planner.cpp(446):

```
#ifdef FAN_SOFT_PWM
fanSpeedSoftPwm = tail_fan_speed;
#else
analogWrite(FAN_PIN, tail_fan_speed);//<<<<<<<<<<
#endif //!FAN_SOFT_PWM
#endif // HAS_FAN
```

So if FAN_SOFT_PWM is defined where/how is it writing fan_pin? which it is since it runs.
there must be another definition for the fan pin and/or another place one of these variables is getting changed
fanspeed, fan, tail_fan_speed, fan_pin

**Comment by thinkyhead** [comment]
> if FAN_SOFT_PWM is defined where/how is it writing fan_pin

It's a software routine. It does what PWM does. It set the pin to 255 or to 0, but in pulses.

**Comment by thinkyhead** [comment]
> I have no servos

Good, that eliminates one possibility.

**Comment by thinkyhead** [comment]
> why does it work when FAN_SOFT_PWM is defined?

Because if something else was - for example - turning the pin off, the soft PWM routine would just keep turning it back on. Not so with the hardware PWM, which (AFAIK) only sets the pin value when it first changes.

**Comment by thinkyhead** [comment]
> another place one of these variables is getting changed: fanspeed, fan, tail_fan_speed, fan_pin

`FAN_PIN` can't change, as it's not a variable. The `fanspeed` is whatever it says on your controller. There's no `fan` variable. The `tail_fan_speed` variable only exists in the logic of determining whether to apply the current fan speed or the fan speed from a printing session. No, something else is going on.

Are you able to attach an oscilloscope to your fan output?

**Comment by ruggb** [comment]
I could - but whatever I would see isn't going to run a fan or turn on an LED. Is there a clue to be had? There is an LED on both outputs so I don't actually need a fan to see an output. If I were getting short pulses that wouldn't run the fan I can see them on the LED. My fan won't run under 20% but the LED blinks.
Let me reiterate the symptoms:
set to EFB the fan works fine on 9
changing fan_pin to 11 (or 6) and setting to EEB causes the fan to only run at 100%. 99% and below is dead except for the kickstart pulse.
defining FAN_SOFT_PWM causes the fan to work fine on 11 (or 6) but the heaters no longer work.
when I clk them on they sit for 5-10 sec then turn off. Not temp rise and I suspect the bed would do the same thing if I had it connected.
I am having a challenge reading this code (I'm not a programmer) but as far as I can tell there is no other 11 or 6 defined.

**Comment by ruggb** [comment]
BTW - re fan_pin, I was referring to something else changing the pin# - I did not mean to imply that it was a "variable".

**Comment by thinkyhead** [comment]
@nophead @daid @boelle Can you can make any sense of this issue? I'm not up on the quirks of overriding pins for the standard RAMPS electronics. Why should Pin 11 with EEB act differently from Pin 9 with EFB?

**Comment by boelle** [comment]
no clue, but i have decided to replace my 644 based electronics to ramps again, but it will be 2 weeks befor that happens

but once i had ramps i did not have any issues, prob because i have a very basic printer. only 1 extruder

**Comment by thinkyhead** [comment]
@boelle I love my RAMPS. P.S. I wonder, can you please come to my defense in [this topic](https://github.com/MarlinFirmware/Marlin/issues/2090)? It's really bumming me out to be dumped on for making an honest attempt.

**Comment by daid** [comment]
@thinkyhead no clue. I've only worked with Ultimaker electronics. (but could be different PMW pins, or some pins not PWMable)

**Comment by boelle** [comment]
all i did then was setting my boards as ramps and it worked... cant
remember the suffix but it was of course the one with bed extruder and fan.

if i should do any guessing a file have been changed that screw up what pin
do what

Den mandag den 18. maj 2015 skrev Scott Lahteine notifications@github.com:

**Comment by thinkyhead** [comment]
> could be different PMW pins, or some pins not PWMable

Yes, similar to my thinking. Perhaps the pin is used for a timer in some case, but not in others. Not sure which timer corresponds to Pin 10-11. I'll have to check and see if the source assigns different timers to different hardware.

**Comment by boelle** [comment]
maybe the user could post config file on pastebin.com and give link here

and maybe the boards pin file

Den mandag den 18. maj 2015 skrev Scott Lahteine notifications@github.com:

**Comment by thinkyhead** [comment]
@boelle Well the board pins part is easy. It's just a `BOARD_RAMPS_EEB` with one single difference – he is overriding the `FAN_PIN` so that it is 11. So all we need to see is the `Configuration.h` (and, maybe `Configuration_adv.h`).

**Comment by boelle** [comment]
yep, my reason for the pins file was that if he have changed that one and somehow made things cross

**Comment by thinkyhead** [comment]
> somehow made things cross

That is my general thinking. EEB might be using Pin 11 for something already – but in disguise, either as a timer, or as a digital –rather than analog– pin. Software PWM works, and that is a clue, but it affects heating. The idea is to turn the 2nd "E" in EEB into an "F." Seems to me it would be better to use EFB and just wire the right pins.

**Comment by boelle** [comment]
hmm both pins at least are PWM pins... ie i was looking at at mega2560 pin map so that far it should be ok... 

next thing is that we have a few flavours or ramps boards but to my knowledge there is only one board out there.... 

btw... pins 6 and 11 are normmaly for the servos.... yes you can connect a fan to them but they cant drive a fan directly... you have to use a transistor as middle man, ie the pin drives the transistor and the transistor drives the fan... a fan direct will take to much current

**Comment by boelle** [comment]
btw... you dont need to change the fan pin either.... 

you just turn on the fan with G-Code... M42 P11 S255 for full on... off you just change S255 to S0

**Comment by thinkyhead** [comment]
> a fan direct will take too much current

Ah, interesting. So, a setting of 100 on Pin 11 _could_ actually be more like 255 on Pin 9…

> pins 6 and 11 are normally for the servos

Yeah, but since no servos are enabled 11 is technically "free for use." But yeah, if the pins have a different voltage range that could be problematic. Meanwhile, the conflict with the heaters is still a little mysterious.

**Comment by boelle** [comment]
not quite so.... PWM is a square wave.... so at the on period it will still take the same current

so S100 is the same on both pins and the same current will flow no matter what pin the fan is connected to.... 

the BIG difference is that D11 and D6 goes straight to the main chip on the mega board, and that one are not made to drive a fan.... 

some PC cpu fans have a 4 pin connector... 2 is for power, 1 is pwm output for sensing the current speed, the last one is a PWM input to tell how fast it should go... pwm out is just used to detect if the fan is blocked, bearings are dead etc...

and a servo pin drains almost no power (close to 0 mA) its just there to tell the electronics inside the servo what angle we want...

to drive a fan from the servo pins it needs to be wired like this: http://i.imgur.com/dYAa5.png

and then use the above M42 command to control the speed.... 

the signal pin on the servo connector would always be chip level voltage... either 3.3V or 5V

**Comment by fmalpartida** [comment]
Just to bring some light into the topic. IN The Arduino library set when you do an analogWritre to a pin that is not associated to a timer, it will manage it a a normal digitalWrite. I can't remember the threshold but if memory serves me right, it is 1.
Therefore, no pwm will be synthesized on pins that don't use a timer or it will have to be all SW. 
Having said all this, for the pwm pin to work it has to be in the Arduino pin mapping space, not fastio.

**Comment by ruggb** [comment]
this is being used with a fan extender bd. That info was in a previous thread. That bd uses pin 11 or 6 since it is mounted on the servo pins. (it also blocks usage of 4 & 5)

The pins 2-13 are the same type of I/O. The fan works on 11 or 6 in PWM mode if FAN_SOFT_PWM is enabled. But that breaks the heaters.

The fan will go on as long as I make it 100%. Does M42 do PWM? How would u make Sli3r use M42 instead of M106? 

Issuing M42 P11 S255 via Repetier host does nothing.

If EEB were already using 11, why would it work 100% and if FAN_SOFT_PWM is enabled work PWM? Any other use should break that also.

What is the purpose of FAN_SOFT_PWM? Without it, it works fine on pin 9. and EFB. So why is it needed for EEB and pin 11?

**Comment by boelle** [comment]
You should not need SOFT pwm at all

Yes M42 P? S? tells what pin and how fast.... P? is the digital pin, it has to be a pwm pin or it will not work. S? is how fast and you can use any number between 0 and 255, where 0 is off and 255 is full on. in slic3r i would disable use of fan... ie in filament settings >>> cooling >>>> remove checkmarks from keep always on and auto cooling

then in printer settings >>> custom  G code 

in the start G code add:

M42 P? S?

and in stop G code:

M42 P? S0

**Comment by ruggb** [comment]
@boelle to be clear, currently FAN_SOFT_PWM is not defined.
fan works @ 100% with m106, nothing less.
M42 does nothing - is that a clue?
that would give me constant full fan, that isn't a perfect solution.
I can;t believe I am the only one configuring a printer like this - in fact I know someone else who can't get his working.

**Comment by boelle** [comment]
You have to type M42 P11 S155

M42 on its own does nothing

**Comment by boelle** [comment]
M106 will be 100%... it just a fan on command

it would be the same as M42 P11 S255

replace 11 with the digital pin number that your fan connects to

**Comment by ruggb** [comment]
No, I believe m106 has parameters. It is what slice uses. M107 has none. Just off. I entered the full m42 cmd multiple times with 11 & 6 & 100 & 255---nothing.

**Comment by boelle** [comment]
well i use M42 P15 S255 with a fan connected to digital pin 15 on my board
and that one works fine

Den mandag den 18. maj 2015 skrev ruggb notifications@github.com:

**Comment by ruggb** [comment]
If u don't have an extender bd u can just connect an led &res to the pin to test - it will handle 40 ma. U will definitely see it flashing under maybe 80℅.

**Comment by ruggb** [comment]
Try m42 on pins 11;6;5;4 --15 is not a pwm pin and the issue isn't 255 it is 100 or250 or 50

**Comment by boelle** [comment]
well i can drive the fan lower than that

but something is not right with your setup... M42 P? S? should always work
as it directly tell what pin to use.

it does not even have to be defined in the pins file

Den mandag den 18. maj 2015 skrev ruggb notifications@github.com:

**Comment by ruggb** [comment]
As I indicated above even m42 p11 s255 does nothing.

**Comment by ruggb** [comment]
R u conf for eeb

**Comment by boelle** [comment]
and as i indicated above something is wrong with your setup

**Comment by ruggb** [comment]
I'll take a fresh copy of 1.0.3 dev and set CoreXY, BOARD_RAMPS_EEB, Dir, fan pin, end stops, quickhome, etc. – the minimum needed to run. Maybe it's a CoreXY or quickhome issue - which is what this started out as.

**Comment by ruggb** [comment]
Well, of course. Why do you think I'm discussing this with you?
Do you have a CoreXY machine and quickhome implemented with two extruders?

**Comment by thinkyhead** [comment]
> CoreXY or quickhome

Those are a huge long shot. Those don't have any effect in the area of pins and PWMs, which happens at a lower level. It's best if you test and collect data about the behavior, see how you can affect it, and then the coders (well, me) will have a good idea what is most likely implicated. Then we can lead you through a process of elimination.

**Comment by ruggb** [comment]
I am trying to do that.........all that I have posted is all that I have RE tests & data. Working on a new clean build.

**Comment by ruggb** [comment]
I dnld the latest dev
u changed this in ramps pins line ~88 to 4 which makes no sense since the extender uses 11 or 6.
  #define FAN_PIN            11 // IO pin. Buffer needed //*\* my change to 11

**Comment by thinkyhead** [comment]
You're doing that hard-to-read thing again. Let me decipher, hold on…

Ok, so you're saying _someone_ (possibly "u") might have screwed up the RAMPS configuration for EEB because the FAN_PIN is 4? Well, I can't say. It's been that way for a long, long time, before I even came onto the project, and this is the first bug report about it that I know of.

**Comment by ruggb** [comment]
sorry my mistake - but if a commercial bd exists why use 4 instead of 11 or 6?

**Comment by ruggb** [comment]
the fan extender bd uses servo 11 & 6 not 5 or 4. I guess the other issue might be - what if it is an EEB using servos? mine is not.

**Comment by thinkyhead** [comment]
@ruggb This thread may be of some interest to you on this subject: https://groups.google.com/forum/#!topic/kossel-clear/jlBOrnjHpeM

From the page:

> the code changes above allow you to use MB34 (EEB), single or dual extruder and deploy 5V fans on the servo pins

…and as we know…

> D9 works with 12V fans

…and…

> Save the D9 connection for any fans that are going to be aimed at the printed material.  D9 is easily controlled by the slicer programs

...and...

> EXTRUDER_0_AUTO_FAN_PIN

The auto fans are for cooling above a certain temperature, and not intended for G-Code control.

**Comment by ruggb** [comment]
ok, this is a new build using a new dnld today and it works the same.
These r ALL my changes. line # is off by 2 b/c I entered a comment at the top of each file and a line.
fan runs full only. both extruders heat up.  //*\* is my marker for changed or uncommented line.
It is a Ramps 1.4, w/fan extender bd, and Arduino Mega 2560 R3

---

```
Configuration.h(46): #define STRING_VERSION_CONFIG_H __DATE__ "05/18/15 " __TIME__ // build date and time //**
Configuration.h(47): #define STRING_CONFIG_H_AUTHOR "Bill Ruggirello" // Who made the changes. //**
Configuration.h(48): #define STRING_SPLASH_LINE1 "CC-BOT E2 " STRING_VERSION // will be shown during bootup in line 1 //**
Configuration.h(69):   #define MOTHERBOARD BOARD_RAMPS_13_EEB //**
Configuration.h(74): #define CUSTOM_MACHINE_NAME "CC-BOT 2E" //**
Configuration.h(78): #define MACHINE_UUID "277494b7-b7c8-4eb1-a653-c2e695c24911" //**
Configuration.h(82): #define EXTRUDERS 2 //**
#define EXTRUDER_OFFSET_X {0.0, -30.00} // (in mm) for each extruder, offset of the hotend on the X axis //**
Configuration.h(144): #define TEMP_SENSOR_0 5 //**
Configuration.h(145): #define TEMP_SENSOR_1 5 //**
Configuration.h(183): #define EXTRUDER_WATTS (12.0*12.0/3.5) //  P=I^2/R //**
Configuration.h(184): #define BED_WATTS (12.0*12.0/1.1)      // P=I^2/R //**
Configuration.h(301): #define COREXY //**
Configuration.h(323): const bool X_MIN_ENDSTOP_INVERTING = true; // set to true to invert the logic of the endstop. //**
Configuration.h(324): const bool Y_MIN_ENDSTOP_INVERTING = true; // set to true to invert the logic of the endstop. //**
Configuration.h(325): const bool Z_MIN_ENDSTOP_INVERTING = true; // set to true to invert the logic of the endstop. //**
Configuration.h(326): const bool X_MAX_ENDSTOP_INVERTING = true; // set to true to invert the logic of the endstop. //**
Configuration.h(327): const bool Y_MAX_ENDSTOP_INVERTING = true; // set to true to invert the logic of the endstop. //**
Configuration.h(328): const bool Z_MAX_ENDSTOP_INVERTING = true; // set to true to invert the logic of the endstop. //**
Configuration.h(329): const bool Z_PROBE_ENDSTOP_INVERTING = true; // set to true to invert the logic of the endstop. //**
Configuration.h(330): #define DISABLE_MAX_ENDSTOPS //**
Configuration.h(362): #define INVERT_Y_DIR false //**
Configuration.h(363): #define INVERT_Z_DIR true //**
Configuration.h(391): #define X_MAX_POS 335 //**
Configuration.h(392): #define Y_MAX_POS 305 //**
Configuration.h(393): #define Z_MAX_POS 295 //**
Configuration.h(564): #define HOMING_FEEDRATE {50*60, 50*60, 8*60, 0}  // set the homing speeds (mm/min) //**
Configuration.h(568): #define DEFAULT_AXIS_STEPS_PER_UNIT   {200,200,800,317.48}  // default steps per unit for Ultimaker //**
Configuration.h(569): #define DEFAULT_MAX_FEEDRATE          {500, 500, 8, 25}    // (mm/sec) //**
Configuration.h(570): #define DEFAULT_MAX_ACCELERATION      {9000,9000,100,10000}    // X, Y, Z, E maximum start speed for accelerated moves. E default values are good for Skeinforge 40+, for older versions raise them a lot. //**
Configuration.h(616): #define PLA_PREHEAT_HOTEND_TEMP 210 //**
Configuration.h(677): #define REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER //**


Configuration_adv.h(97): #define FAN_KICKSTART_TIME 100 //**
Configuration_adv.h(213): #define HOMING_BUMP_DIVISOR {8, 8, 4}  // Re-Bump Speed Divisor (Divides the Homing Feedrate) //**
Configuration_adv.h(214): #define QUICK_HOME  //if this is defined, if both x and y are to be homed, a diagonal move will be performed initially. //**

pins_RAMPS_13.h(89):   #define FAN_PIN            11 // IO pin. Buffer needed //**
```

**Comment by ruggb** [comment]
I believe I saw that article a while ago. It appeared to me that anyone who got the fan working did so on D9 with mb 33 defined. I do not see any dual extruder mb 34 (EEB) running a fan on a servo pin - or any other pin for that matter. EXCEPT maybe extruder fans and controller fans, which r not what I need (yet). But I don't even think those r running with a dual extruders.
I want a fan that Slic3r program controls with M106 while running 2 extruders and a bed.
That is clearly what they sell this expander bd for. It says so in the literature.

**Comment by thinkyhead** [comment]
What is the part number of the extender board? Is BD the brand name?

**Comment by ruggb** [comment]
here is a link
http://www.geeetech.com/wiki/index.php/Reprap_Ramps1.4_RRD_Fan_Extender

**Comment by ruggb** [comment]
no BD is short for board

**Comment by thinkyhead** [comment]
What happens if you leave all pins as they are, but use `#define EXTRUDER_0_AUTO_FAN_PIN 11` and `#define EXTRUDER_AUTO_FAN_SPEED 255`? (Extruder fans turn on when the extruder heats up.)

**Comment by ruggb** [comment]
AND THE ANSWER IS.........
I found this review:
works great. Needed to find some documentation on the web but did not have too many problems finding it. Control is on D6/D11 as indicated so Marlin board type 34 does not work directly, (Marlin Type 34 controls on D4). But it can be installed shifted two rows of pins towards Aux-1 and will then be on on D4/D5. 

So, I moved the FE over 2 connections so 4 & 5 would be connected instead of 6 & 11.
left everything default:
EEB, quickhome, 2E, temp, fan_pin,.............
WORKS PERFECTLY.
When I put everything back and chg fan_pin to 11 it does not work
chg to 6 - works
chg to 5 - moved fe bd over 2 slots - works
chg to 4 - works

so for some reason pin 11 does not work.
The question is: Is it Arduino or Marlin?
If it is Arduino, it is a defect. HOWEVER, why was Marlin set at 4, is there a conflict with 11 maybe with servo 0?
They appeared to have shifted it b/c Fan_Pin is set to 4 default instead of chg to 11 or 6.
I thought I set it to 6 at one time in the past, but maybe not.

thx

**Comment by nophead** [comment]
It's probably because the different PWM capable pins use different timers
and Marlin needs some timers for its interrupts.

**Comment by ruggb** [comment]
yeah, I read that but it did not involve pin 11 or pin 4
"The PWM outputs generated on pins 5 and 6 will have higher-than-expected duty cycles. This is because of interactions with the millis() and delay() functions, which share the same internal timer used to generate those PWM outputs. This will be noticed mostly on low duty-cycle settings (e.g 0 - 10) and may result in a value of 0 not fully turning off the output on pins 5 and 6. "

**Comment by nophead** [comment]
Pin 4 uses timer 0
Pin 11 uses timer 1
Pin 5 uses timer 3
Pin 6 uses timer 4

Timer 1 is used for the stepper interrupt.

**Comment by ruggb** [comment]
where can I find that info
thx

**Comment by nophead** [comment]
http://www.arduino.cc/en/uploads/Hacking/PinMap2560big.png

**Comment by ruggb** [comment]
perfect - I may have something like that but where is the info as to what the pins r used for. Like how to know which pins use which timers or is that part of the name of the pins?

**Comment by nophead** [comment]
For full details see the datasheet for the chip but the simple answer is digital pin 11 is labelled OC1A and the 1 means timer 1.

**Comment by ruggb** [comment]
thx - I saw that labeling someplace but did not read it carefully then. got to find it again

**Comment by ruggb** [comment]
maybe this needs another thread
I issue the command M303 S210. Temp indicator for E0 rises but E0 & E1 still shows OFF. I get the parameters and it is finished.
I issue the command M303 E1 S210. Temp indicator for E0 rises but E0 & E1 still shows OFF. I get the parameters and it is finished.
If I set E0 ON, E0 therm rises. If I set E1 ON, E1 therm rises.
Why doesn't it show the heater as on during the test?
Why doesn't it show the correct therm rising with an M303 command?

**Comment by ruggb** [comment]
more... I'm using Repetier Host: when E1 only is on no temp curve is shown. When on, the extruders only seem to get to ~ 150°C as measured with an IR meter and the filament doesn't flow, even though the indication is 210. I am using 5 for the therm on an E3D ver 6 clone.

**Comment by thinkyhead** [comment]
`M303` doesn't update the LCD or send temperature info to the host during the test. That's ok.

**Comment by thinkyhead** [comment]
> when E1 only is on no temp curve is shown

You should definitely request this feature on the Repetier Host project issue queue.

**Comment by thinkyhead** [comment]
> even though the indication is 210

Bad thermistor? Or maybe it's not "5" but "1." Try some other values and see if it matches the IR meter better.

**Comment by nophead** [comment]
You can't get an accurate reading from a hot end with IR. It is too small
and the emissivity of metals is too low so it will always read low. A
thermocouple placed inside the nozzle is the only way to get an accurate
reading.

**Comment by ruggb** [comment]
thx
as it turns out I have a thermocouple for this meter. I found another thermistor file. It is linearly about 4°C off now, both nozzles. The IR actually agrees pretty much with the thermocouple within a few degrees. It is all about where u aim it and what the surface is.
There is another table I got from the Ebay listing for an E3D v3 clone. It has a comment about a QU-BD silicone bed with the same #. #13 was close, but this one seems better. U might want to include it. I replaced #8 as that was the way they made it.

```
// for an NTC 3950 as supplied on an E3D V3 Clone 
#if (THERMISTORHEATER_0 == 8) || (THERMISTORHEATER_1 == 8) || (THERMISTORHEATER_2 == 8) || (THERMISTORBED == 8) // QU-BD silicone bed QWG-104F-3950 thermistor
const short temptable_8[][2] PROGMEM = {
         {1*OVERSAMPLENR,        938},
         {31*OVERSAMPLENR,       314},
         {41*OVERSAMPLENR,       290},
         {51*OVERSAMPLENR,       272},
         {61*OVERSAMPLENR,       258},
         {71*OVERSAMPLENR,       247},
         {81*OVERSAMPLENR,       237},
         {91*OVERSAMPLENR,       229},
         {101*OVERSAMPLENR,      221},
         {111*OVERSAMPLENR,      215},
         {121*OVERSAMPLENR,      209},
         {131*OVERSAMPLENR,      204},
         {141*OVERSAMPLENR,      199},
         {151*OVERSAMPLENR,      195},
         {161*OVERSAMPLENR,      190},
         {171*OVERSAMPLENR,      187},
         {181*OVERSAMPLENR,      183},
         {191*OVERSAMPLENR,      179},
         {201*OVERSAMPLENR,      176},
         {221*OVERSAMPLENR,      170},
         {241*OVERSAMPLENR,      165},
         {261*OVERSAMPLENR,      160},
         {281*OVERSAMPLENR,      155},
         {301*OVERSAMPLENR,      150},
         {331*OVERSAMPLENR,      144},
         {361*OVERSAMPLENR,      139},
         {391*OVERSAMPLENR,      133},
         {421*OVERSAMPLENR,      128},
         {451*OVERSAMPLENR,      123},
         {491*OVERSAMPLENR,      117},
         {531*OVERSAMPLENR,      111},
         {571*OVERSAMPLENR,      105},
         {611*OVERSAMPLENR,      100},
         {641*OVERSAMPLENR,      95},
         {681*OVERSAMPLENR,      90},
         {711*OVERSAMPLENR,      85},
         {751*OVERSAMPLENR,      79},
         {791*OVERSAMPLENR,      72},
         {811*OVERSAMPLENR,      69},
         {831*OVERSAMPLENR,      65},
         {871*OVERSAMPLENR,      57},
         {881*OVERSAMPLENR,      55},
         {901*OVERSAMPLENR,      51},
         {921*OVERSAMPLENR,      45},
         {941*OVERSAMPLENR,      39},
         {971*OVERSAMPLENR,      28},
         {981*OVERSAMPLENR,      23},
         {991*OVERSAMPLENR,      17},
         {1001*OVERSAMPLENR,     9},
         {1021*OVERSAMPLENR,     -27}
};

#endif 
```

**Comment by boelle** [comment]
will close this one.... not 100% sure that there ever was an issue with marlin

**Comment by FalloutBe** [comment]
Hello.

I followed ruggb's instructins and it helped me out a lot to get my FE to work. I can now control both my part cooling fans and my LEDs by using M42 (or the slider in repetier for the fans)

The problem however is: #define FAST_PWM_FAN is no longer working.
This does work when using D6 and D11, but when using D4 and D5 it does not.

I can't use D6 and D11 because then D11 for some reason only goes 0 or 255 but not pwm values in between.
Any ideas?

**Comment by ruggb** [comment]
what is the reason u want FAST_PWM_FAN defined?
Mine is not defined.
I assume u r driving ur fans via 4 & 5 as I am. - actually only using 1 of them but both work fine.
I do have the kickstart defined. My fans do not start below 35. they don't run under 20 either.
#define FAN_KICKSTART_TIME 100

**Comment by FalloutBe** [comment]
Oh thanks for the really quick reply! I thought it would take a while since this thread was closed.

Well, I use D4 for my part cooling fans and D5 for my LED lighting.
Both are really annoying on the low switching frequency. The fans make a LOT of noise and the LEDs are flickering which also annoys me ^^

No, I did not have kickstart defined, didn't even know this existed but I will certainly use it!

**Comment by ruggb** [comment]
OK, I don't remember if I ever had that defined, but my part fans r quite enough and my LED strip is hot wired full on all the time, as are my hotend fans so I guess u might want to start another issue to find out. It may be that it only applies to the std fan pins, but I would expect it to apply at least to whatever pin is defined for the fan.
Can't help u further.

**Comment by ruggb** [comment]
OH BTW, what did u configure to drive ur LEDs from D5? and how do u set them on?

**Comment by FalloutBe** [comment]
No problem, I will continue using the low frequency PWM. At least it is working now!

I used the "scripted buttons" in Repetier-Host. They send the M42 command to turn them on.
I have a button for 100%, one for 50% and one for 0%.
Sadly I can't seem to find how I can add a slider like the fan slider, to control the LEDs :/

Might it not be possible to trick Repetier into thinking there is a second fan somewhere, so that I would be granted an additional slider? ^^

**Comment by ruggb** [comment]
U might find that on the Repetier site

**Comment by nicksears** [comment]
For those following up on the original question. As of Jan 15, 2016, with marlin 1.1.0, there's still an error in "pins_RAMPS_13.h" for a ramps 1.4/1.4 setup using a geeetech rrd fan extender. 

The fan section neds to be modified (likely at the very end) to say "#define FAN_PIN 6 // IO pin. Buffer needed" instead of using pin 4. 

Then pin 6 will be associated with m106, and pin 11 will be controllable with "M42 P11 SXXX"

**Comment by ruggb** [comment]
I had to shift the bd over 2 pins and use pin 4 or 5. I didn't pursue figuring out why, just assumed it was some conflict. I have a Ramps 1.4 EEB - so this works for me.

**Comment by Wackerbarth** [comment]
I suspect that there is no single "correct" way to do things with only one pins file for the general class of EEB configurations.  We really need to work on developing a wiring HAL that customizes the firmware for a particular wiring schematic. Unless the configuration is tied to an actual wiring diagram, there will always be a disagreement as to the "correct" set of definitions.

**Comment by nicksears** [comment]
That doesn't apply on this situation, though. If you have an EEB setup, you
have no fan unless you use the red fan extender, and therefore need it to
be pin 6. There is no other common setup (with a breadboard) and this
should be fixed immediately and permanently. At the very least, there
should be an EEBFE variant.

**Comment by ruggb** [comment]
@nicksears = that fan extender will mount shifted over 2 pins
when u do that u r connected to i/o pin 4 & 5
both of those work for me.
I could not get it to work in the "normal" position on pin 6 or 11.

**Comment by nicksears** [comment]
I downloaded the latest release of marlin 1.1.0 on Jan 15, 2016. With a
standard ramps 1.4/arduino mega with a rrd fan extender/rrd smart
controller lcd this works perfectly when plugged in as intended, but only
if that fan pin is changed to pin 6. Is there any reason to NOT redefine
the pin? Plugging things in incorrectly is the wrong way to solve this
issue.

**Comment by KiteLab** [comment]
@nicksears 
Yes. The fan extender blocks two more pins than necessary when mounted as designed. Mounting it shifted leaves 2 pins open - you can use for servos - or what else.
So mounting it **incorrect** is the far better way to mount it.

**Comment by nicksears** [comment]
OK, I stand corrected, I'll suggest that change on the documentation after
I give it a try.

**Comment by nicksears** [comment]
This seems to work. An extra note, if you change the pin to 6, you're using
pin 6 for the fan_pin and pin 11 for the extra configurable output (work
light for me). If you shift it two pins (away from the power connectors,
the only direction you can go unless you have no reset button) then you are
keeping pin 4 as the fan_pin and pin 5 will be the extra configurable
output. Then you have the servo outputs pin 6 and 11 for ramps 1.4 (6 and 7
for 1.3). I've emailed geeetech with the suggested changes for the wiki.

On Sat, Jan 16, 2016 at 1:30 PM, Nick Sears nicksears@gmail.com wrote:

> OK, I stand corrected, I'll suggest that change on the documentation after
> I give it a try.
> On Jan 16, 2016 1:21 PM, "KiteLab" notifications@github.com wrote:
> 
> > @nicksears https://github.com/nicksears
> > Yes. The fan extender blocks two more pins than necessary when mounted as
> > designed. Mounting it shifted leaves 2 pins open - you can use for servos -
> > or what else.
> > So mounting it _incorrect_ is the far better way to mount it.
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/MarlinFirmware/Marlin/issues/2104#issuecomment-172247977
> > .

##

**Comment by fixbroke** [comment]
I know this is a pretty dead conversation, but I thought I'd add my findings after this thread helped me with my fan extender issues. Hopefully anyone else having this problem can find this thread and use this info.

Like ruggb, I was having problems while trying to use pin D11 for PWM on a print fan. While the fan was on, my steppers would freak out and not work correctly, running in stuttering movements with dropped steps aplenty. This problem didn't occur if the fan was on full blast - only if it was set at a reduced speed. The other fan I am trying to run on the extender is a hotend fan that turns on when the hotend reaches a desired temperature (auto_fan, as it's called in Marlin).

One solution seems to be moving the fan extender over two pins, as ruggb did, to use pins 4 and 5 (instead of 6 and 11). 

I used another solution. In my case, since my auto_fan runs at full speed or not at all (as is the default setting in Marlin), I simply swapped the connectors for my print fan and auto_fan so the print fan used D6 and the auto_fan used pin D11. This way, there is no PWM conflict with the stepper motors since D11 is not being used as a PWM signal.

There is another potential problem that some people may have with D11 if their extender board was purchased cheaply on eBay or AliExpress. My board and some others have a manufacturing defect that prevents the positive pin for the D11 output from ever receiving power. The negative pin is switched by the hardware, but the positive pin is dead. The solution in this case is to solder a connection between the extender board's +12V power supply and the extender board's positive D11 output.
